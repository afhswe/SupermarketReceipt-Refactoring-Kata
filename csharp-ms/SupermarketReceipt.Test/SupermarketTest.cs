using System.Collections.Generic;
using System.Text.Json;
using ApprovalTests.Combinations;
using ApprovalTests.Reporters;
using System.Text.Json.Serialization;
using Xunit;
using FluentAssertions;
using Newtonsoft.Json;
using JsonSerializer = System.Text.Json.JsonSerializer;

namespace SupermarketReceipt.Test
{
    public class SupermarketTest
    {
        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void ApproveReceiptCalculation()
        {
            var itemQuantities = new double[]
            {
                1,
                5
            };
            var productUnits = new[]
            {
                ProductUnit.Each,
                ProductUnit.Kilo
            };
            var specialOfferTypes = new[]
            {
                SpecialOfferType.TenPercentDiscount,
                SpecialOfferType.FiveForAmount,
                SpecialOfferType.ThreeForTwo,
                SpecialOfferType.TwoForAmount
            };
            var specialOfferArguments = new[]
            {
                10.0
            };

            CombinationApprovals.VerifyAllCombinations(
                CallCheckoutAndGetReceipt,
                itemQuantities,
                productUnits,
                specialOfferTypes,
                specialOfferArguments);
        }

        private string CallCheckoutAndGetReceipt(
            double itemQuantity,
            ProductUnit productUnit,
            SpecialOfferType specialOfferType,
            double specialOfferArgument)
        {
            var catalog = new FakeCatalog();
            var cart = new ShoppingCart();
            var teller = new Teller(catalog);

            var product = new Product($"chocolate", productUnit);
            catalog.AddProduct(product, 1.00);
            cart.AddItem(product);
            cart.AddItemQuantity(product, itemQuantity);
            teller.AddSpecialOffer(specialOfferType, product, specialOfferArgument);

            var receipt = teller.CheckOutArticlesFrom(cart);
            var receiptAsJson = ReceiptAsJson(receipt);
            return receiptAsJson;
        }

        private string ReceiptAsJson(Receipt receipt)
        {
            var receiptForTest = new ReceiptDataForPrinting();
            receiptForTest.Discounts = receipt.GetDiscounts();
            receiptForTest.TotalPrice = receipt.GetTotalPrice();
            receiptForTest.Items = receipt.GetItems();
            receiptForTest.SpecialOffer = SpecialOfferType.TenPercentDiscount;
            return JsonSerializer.Serialize(
                receiptForTest,
                new JsonSerializerOptions()
                {
                    WriteIndented = true,
                    Converters =
                    {
                        new JsonStringEnumConverter()
                    }
                });
        }
    }

    public class ReceiptDataForPrinting
    {
        public double TotalPrice { get; set; }
        public SpecialOfferType SpecialOffer { get; set; }
        public List<Discount> Discounts { get; set; }
        public List<ReceiptItem> Items { get; set; }
    }
}