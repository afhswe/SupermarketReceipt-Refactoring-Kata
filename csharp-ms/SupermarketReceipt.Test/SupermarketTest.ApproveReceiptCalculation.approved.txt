[1,Each,TenPercentDiscount,10] => {
  "TotalPrice": 1.8,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "10% off",
      "DiscountAmount": -0.2,
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    }
  ]
}
[1,Each,FiveForAmount,10] => {
  "TotalPrice": 2,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    }
  ]
}
[1,Each,ThreeForTwo,10] => {
  "TotalPrice": 2,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    }
  ]
}
[1,Each,TwoForAmount,10] => {
  "TotalPrice": 10,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "2 for 10",
      "DiscountAmount": 8,
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    }
  ]
}
[1,Kilo,TenPercentDiscount,10] => {
  "TotalPrice": 1.8,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "10% off",
      "DiscountAmount": -0.2,
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    }
  ]
}
[1,Kilo,FiveForAmount,10] => {
  "TotalPrice": 2,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    }
  ]
}
[1,Kilo,ThreeForTwo,10] => {
  "TotalPrice": 2,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    }
  ]
}
[1,Kilo,TwoForAmount,10] => {
  "TotalPrice": 10,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "2 for 10",
      "DiscountAmount": 8,
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    }
  ]
}
[5,Each,TenPercentDiscount,10] => {
  "TotalPrice": 5.4,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "10% off",
      "DiscountAmount": -0.6,
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 5,
      "Quantity": 5
    }
  ]
}
[5,Each,FiveForAmount,10] => {
  "TotalPrice": 11,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "5 for 10",
      "DiscountAmount": 5,
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 5,
      "Quantity": 5
    }
  ]
}
[5,Each,ThreeForTwo,10] => {
  "TotalPrice": 4,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "3 for 2",
      "DiscountAmount": -2,
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 5,
      "Quantity": 5
    }
  ]
}
[5,Each,TwoForAmount,10] => {
  "TotalPrice": 30,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "2 for 10",
      "DiscountAmount": 24,
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Each"
      },
      "Price": 1,
      "TotalPrice": 5,
      "Quantity": 5
    }
  ]
}
[5,Kilo,TenPercentDiscount,10] => {
  "TotalPrice": 5.4,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "10% off",
      "DiscountAmount": -0.6,
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 5,
      "Quantity": 5
    }
  ]
}
[5,Kilo,FiveForAmount,10] => {
  "TotalPrice": 11,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "5 for 10",
      "DiscountAmount": 5,
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 5,
      "Quantity": 5
    }
  ]
}
[5,Kilo,ThreeForTwo,10] => {
  "TotalPrice": 4,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "3 for 2",
      "DiscountAmount": -2,
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 5,
      "Quantity": 5
    }
  ]
}
[5,Kilo,TwoForAmount,10] => {
  "TotalPrice": 30,
  "SpecialOffer": "TenPercentDiscount",
  "Discounts": [
    {
      "Description": "2 for 10",
      "DiscountAmount": 24,
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      }
    }
  ],
  "Items": [
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 1,
      "Quantity": 1
    },
    {
      "Product": {
        "Name": "chocolate",
        "Unit": "Kilo"
      },
      "Price": 1,
      "TotalPrice": 5,
      "Quantity": 5
    }
  ]
}
